presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java

```diff
@@ -0,0 +1,307 @@
+
+            if (callName.equals("AND") || callName.equals("OR")) {
+                // return sorted distinct arguments
+                Set<RowExpression> distinctArguments = ImmutableSet.copyOf(call.getArguments());
+                if (distinctArguments.size() == 1) {
+                    return Iterables.getOnlyElement(distinctArguments);
+                }
+
+                return new CallExpression(
+                        call.getSignature(),
+                        call.getType(),
+                        ROW_EXPRESSION_ORDERING.sortedCopy(distinctArguments));
+            }
+
+            if (callName.equals(mangleOperatorName(EQUAL)) || callName.equals(mangleOperatorName(NOT_EQUAL)) || callName.equals("IS_DISTINCT_FROM")) {
+                // sort arguments
+                return new CallExpression(
+                        call.getSignature(),
+                        call.getType(),
+                        ROW_EXPRESSION_ORDERING.sortedCopy(call.getArguments()));
+            }
+
+            if (callName.equals(mangleOperatorName(GREATER_THAN)) || callName.equals(mangleOperatorName(GREATER_THAN_OR_EQUAL))) {
+                // convert greater than to less than
+                return new CallExpression(
+                        new Signature(
+                                callName.equals(mangleOperatorName(GREATER_THAN)) ? mangleOperatorName(LESS_THAN) : mangleOperatorName(LESS_THAN_OR_EQUAL),
+                                SCALAR,
+                                call.getSignature().getTypeParameterRequirements(),
+                                call.getSignature().getReturnType(),
+                                swapPair(call.getSignature().getArgumentTypes()),
+                                false),
+                        call.getType(),
+                        swapPair(call.getArguments()));
+            }
+
+            return call;
+        }
+
+        @Override
+        public RowExpression visitConstant(ConstantExpression constant, Void context)
+        {
+            return constant;
+        }
+
+        @Override
+        public RowExpression visitInputReference(InputReferenceExpression node, Void context)
+        {
+            return node;
+        }
+    }
+
+    private static class RowExpressionComparator
+            implements Comparator<RowExpression>
+    {
+        private final Comparator<Object> classComparator = Ordering.arbitrary();
+        private final ListComparator<RowExpression> argumentComparator = new ListComparator<>(this);
+
+        @Override
+        public int compare(RowExpression left, RowExpression right)
+        {
+            int result = classComparator.compare(left.getClass(), right.getClass());
+            if (result != 0) {
+                return result;
+            }
+
+            if (left instanceof CallExpression) {
+                CallExpression leftCall = (CallExpression) left;
+                CallExpression rightCall = (CallExpression) right;
```

Comment:
I'm not sure this is right. i don't think we can assume that the right is a CallExpression if the left is a CallExpression? What is guaranteeing that it is not something like a ConstantExpression?